### PR TITLE
adding UniFunc recipe

### PR DESCRIPTION
**Pull request by PedroMTQ**
Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [ ] If static libraries are linked in, the license of the static library is packaged.
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [ ] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

**Comment by PedroMTQ** [comment]
@conda-forge/core
This is failing at "run docker build". could I get some help?

**Comment by PedroMTQ** [comment]
@conda-forge/core  This is failing at "run docker build". could I get some help?

**Comment by BastianZim** [comment]
The `1.2` release doesn't have a setup.py or pyproject.toml which is why it fails: https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=432739&view=logs&j=6f142865-96c3-535c-b7ea-873d86b887bd&t=22b0682d-ab9e-55d7-9c79-49f3c3ba4823&l=1123

Are you able to normally install this package outside of conda-forge?

**Comment by PedroMTQ** [comment]
> The `1.2` release doesn't have a setup.py or pyproject.toml which is why it fails: https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=432739&view=logs&j=6f142865-96c3-535c-b7ea-873d86b887bd&t=22b0682d-ab9e-55d7-9c79-49f3c3ba4823&l=1123
> 
> Are you able to normally install this package outside of conda-forge?

I added the setup.py.

This is my first time packaging something so I'm not entirely sure if I did everything correctly. Could you please confirm and let me know?

Regards,
Pedro

File: recipes/UniFunc/meta.yaml
Comment (BastianZim) [inline recipes/UniFunc/meta.yaml:1]:
```suggestion
```

File: recipes/UniFunc/meta.yaml
Comment (BastianZim) [inline recipes/UniFunc/meta.yaml:4]:
```suggestion


```

File: recipes/UniFunc/meta.yaml
Comment (BastianZim) [inline recipes/UniFunc/meta.yaml:10]:
```suggestion
  url: https://github.com/PedroMTQ/UniFunc/archive/refs/tags/{{ version }}.tar.gz
```
This is sufficient, the name shouldn't change so no placeholder required.

File: recipes/UniFunc/meta.yaml
Comment (BastianZim) [inline recipes/UniFunc/meta.yaml:12]:
```suggestion
```
Was this needed? Normally, condo-build should find it automatically.

File: recipes/UniFunc/meta.yaml
Comment (BastianZim) [inline recipes/UniFunc/meta.yaml:15]:
```suggestion
  number: 0
  noarch: python
  script: {{ PYTHON }} -m pip install . -vv
```
This installs your package.

File: recipes/UniFunc/meta.yaml
Comment (BastianZim) [inline recipes/UniFunc/meta.yaml:21]:
```suggestion
  build:
    - setuptools >=42
    - wheel
  host:
    - pip
    - python >=3.6
```
I assumed that `python >=3.6` is your minimum version, you might need to change that.

File: recipes/UniFunc/meta.yaml
Comment (BastianZim) [inline recipes/UniFunc/meta.yaml:26]:
```suggestion
  run:
    - nltk
    - numpy
    - python
    - requests
```
You don't have any requirements listed in your _setup.py_ or _pyprpoject.tom_.
Are these all? Might be good to add them to your project as well.

File: recipes/UniFunc/meta.yaml
Comment (BastianZim) [inline recipes/UniFunc/meta.yaml:30]:
```suggestion
test:
  imports:
    - unifunc
  commands:
    - pip check
  requires:
    - pip
```
If you list the requirements in your _setup.py_, this should make sure that you're not missing any here.
Do you have any tests that you run upstream?

File: recipes/UniFunc/meta.yaml
Comment (BastianZim) [inline recipes/UniFunc/meta.yaml:37]:
```suggestion
  summary: Tool for similarity analysis of protein function annotations.
```
This is assumed.

File: recipes/UniFunc/meta.yaml
Comment (BastianZim) [inline recipes/UniFunc/meta.yaml:42]:
```suggestion
  description: |
    UniFunc is a text mining tool that processes and analysis text similarity
    between a pair of protein function annotations. It is mainly used as a
    cross-linking mechanism or redundancy elimination tool when processing
    annotations without any sort of database identifiers.
```

File: recipes/UniFunc/meta.yaml
Comment (BastianZim) [inline recipes/UniFunc/meta.yaml:47]:
```suggestion
extra:
  recipe-maintainers:
    - PedroMTQ
```

**Review by BastianZim — CHANGES_REQUESTED** [review]
Looks good just a couple of small style guide and indentation changes.

From what I can see, the _setup.py_ and _pyproject.toml_ are not yet included in the 1.2 release. I would suggest cutting a new release and then updating the recipe to use that.

File: recipes/UniFunc/meta.yaml
Comment (PedroMTQ) [reply recipes/UniFunc/meta.yaml]: I have the setup.py here:
https://github.com/PedroMTQ/UniFunc/blob/main/setup.py

I think it only needs setuptools to run, I assume packages to run the actual tool are not included here?

File: recipes/UniFunc/meta.yaml
Comment (BastianZim) [inline recipes/UniFunc/meta.yaml:28]:
```suggestion
    - python >=3.6
```
Sorry, forgot that.

**Comment by PedroMTQ** [comment]
@BastianZim 
I created a new release with the setup.py and pyproject.toml. I'm not sure if they are ok though

File: recipes/UniFunc/meta.yaml
Comment (BastianZim) [reply recipes/UniFunc/meta.yaml]: Packages that are listed in `install_requires` are required when running the package.
`setuptools >=42` would therefore be in `setup_requires` and `nltk` etc. in `install_requires`.

Ref: https://packaging.python.org/en/latest/tutorials/packaging-projects/

**Comment by BastianZim** [comment]
Lets see. It might be good to add the `install_requires` section from https://github.com/conda-forge/staged-recipes/pull/17356#discussion_r775856495

The link contains the python approach to packaging so it should also give some useful general tips for the rest of your package.

File: recipes/UniFunc/meta.yaml
Comment (BastianZim) [inline recipes/UniFunc/meta.yaml:12]: This needs to be updated to the newest one.

Comment (PedroMTQ) [reply recipes/UniFunc/meta.yaml]: done

**Comment by PedroMTQ** [comment]
> `install_requires`

Here I would just add `setuptools>=42?` and `wheel`?

**Comment by PedroMTQ** [comment]
@BastianZim 
This is now failing at the tests
Any idea why the `import unifunc` is not working?

**Comment by BastianZim** [comment]
`install_requires` is what gets installs with your package when you run it so it's the equivalent of `run` here.
`setup_requires` is where setup tools etc. should be added but you already have that in your pyproject.toml.

> tests

It might actually be that the import is `UniFunc`. I changed it because of the package name here but let's try it with that.

Generally, it's good practice not make sure that you can build your package locally first using the GitHub repo and pip (Which can install a package directly from GitHub. Once that is working, conda should also work, as it also just calls pip during the build step.

**Comment by PedroMTQ** [comment]
@BastianZim 
I tried this again with the corrected import and it still fails.
I'm not sure why this fails. Do I need to restructure the repo in a different manner?
I tried it locally, outside and inside the UniFunc folder and I manage to import it using `import UniFunc`.
I'm not sure why it fails with conda?

I didn't try locally with pip before, I wasn't aware that's how I'm supposed to do it.

**Comment by BastianZim** [comment]
It also fails if I import it locally so something in your setup.py is missing. From a cursory glance, you're not defining what your package actually is (Folder where the package lives etc.). This is required as otherwise `pip` doesn't know which files to actually install.

I would suggest updating your repo to a point where `pip install https://github.com/PedroMTQ/UniFunc/archive/refs/tags/1.2.1.tar.gz` works and once that is working it should be very simple to add it here. 
There are many guides on packaging python like this one that will show you how to do this: https://realpython.com/pypi-publish-python-package/
Since your package is primarily command line based it might also make sense to add entry points: https://packaging.python.org/en/latest/specifications/entry-points/

> I didn't try locally with pip before, I wasn't aware that's how I'm supposed to do it.

No worries, just makes it easier as you don't need to wait for the CI every time.

By the way, the python dependency should be separate (in your _setup.py_) under `python_requires`.

**Comment by PedroMTQ** [comment]
@BastianZim 

I have updated the repo and the recipe and it seems all checks passed now.
What happens now? Do I need to do anything else for it the package to be in conda forge?

\Pedro

**Comment by PedroMTQ** [comment]
@conda-forge/help-python The recipe has passed all checks and is ready for your review. Thanks.

**Comment by BastianZim** [comment]
I'm away from my computer so I cannot review it right now but let me ping the team for you so someone can help you. @conda-forge/help-python

File: recipes/UniFunc/meta.yaml
Comment (dopplershift) [inline recipes/UniFunc/meta.yaml:17]: This package looks like it's pure Python, any reason it couldn't be noarch? We can ignore any Windows failures due to missing dependencies:
```suggestion
  noarch: python
```

File: recipes/UniFunc/meta.yaml
Comment (dopplershift) [inline recipes/UniFunc/meta.yaml:24]: These should all be host. Also add Python version for noarch:
```suggestion
  host:
    - setuptools >=42
    - wheel
    - pip
    - python >=3.6
```

File: recipes/UniFunc/meta.yaml
Comment (dopplershift) [inline recipes/UniFunc/meta.yaml:28]: For noarch:
```suggestion
    - python >=3.6
```

File: recipes/UniFunc/meta.yaml
Comment (PedroMTQ) [reply recipes/UniFunc/meta.yaml:17]: I tried with noarch and it was failing on windows.
see here:
https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=437046&view=logs&j=240f1fee-52bc-5498-a14a-8361bde76ba0&t=06d62fd2-ab49-5add-9ee5-83b6882d7d2b

Comment (dopplershift) [reply recipes/UniFunc/meta.yaml:17]: It's failing on Windows because it can't read the `README.md` file with Python's default encoding on Windows ('cp1252'). Either remove the unicode quotes in the document (there are several quotes in there which are the "fancy" quotes) or in `setup.py` call `read_text` explicitly with a utf-8 encoding, like:
```python
read_text(encoding='utf-8')
```

Comment (PedroMTQ) [reply recipes/UniFunc/meta.yaml:17]: ah ok. I  will apply your proposed changes tomorrow
thanks for the help!

**Comment by PedroMTQ** [comment]
@dopplershift 
This is now crashing on windows again in build recipe:
`INFO:conda_build.build:Packaging unifunc
Packaging unifunc
INFO conda_build.build:build(2289): Packaging unifunc
INFO:conda_build.build:Packaging unifunc-1.3.4-pyhd8ed1ab_0
Packaging unifunc-1.3.4-pyhd8ed1ab_0
INFO conda_build.build:bundle_conda(1529): Packaging unifunc-1.3.4-pyhd8ed1ab_0
number of files: 25
Unknown format
Unknown format
Unknown format
Unknown format
Unknown format
Unknown format
Unknown format
Unknown format
Unknown format
Unknown format
Unknown format
Unknown format
Unknown format
Unknown format
Unknown format
Unknown format
   INFO: sysroot: 'C:/Windows/' files: '['winhlp32.exe', 'win.ini', 'twain_32/wiatwain.ds', 'twain_32.dll']'
   INFO (unifunc,Scripts/unifunc.exe): Needed DSO C:/Windows/System32/KERNEL32.dll found in $SYSROOT
   INFO (unifunc,Scripts/unifunc.exe): Needed DSO C:/Windows/System32/SHLWAPI.dll found in $SYSROOT
   INFO (unifunc): Interpreted package 'unifunc' is interpreted by 'python'
[noarch_python] Noarch package contains binary script: unifunc.exe
##[error]Cmd.exe exited with code '1'.
Finishing: Build recipe
`

File: recipes/UniFunc/meta.yaml
Comment (dopplershift) [inline recipes/UniFunc/meta.yaml:17]: Need to add an [entry_point](https://conda-forge.org/docs/maintainer/knowledge_base.html#noarch-builds):
```suggestion
  noarch: python
  entry_points:
    - unifunc=unifunc.__main__:main

```

**Review by dopplershift — APPROVED** [review]

**Comment by PedroMTQ** [comment]
@dopplershift 
Thank you for approving the repo.
Just a question.If I were to update the development repo and release a new version, would I then just need to update the unifunc-feedstock meta.yaml?

Regards,
Pedro

**Comment by BastianZim** [comment]
The conda-forge bots will pick that up and create a PR automatically. You only need to check if any of the dependencies changed. 
But yes, you can also update everything manually by updating the meta.yaml. Just make sure that you’re doing it from a personal fork using a pull request and not in the repo directly.

**Comment by PedroMTQ** [comment]
> The conda-forge bots will pick that up and create a PR automatically. You only need to check if any of the dependencies changed. But yes, you can also update everything manually by updating the meta.yaml. Just make sure that you’re doing it from a personal fork using a pull request and not in the repo directly.

Ah I see, when you mean the bots will pick it up, do you mean when I create a new release in the github repo? (https://github.com/PedroMTQ/unifunc)

**Comment by BastianZim** [comment]
Yea. Will take a couple of hours though.

**Comment by PedroMTQ** [comment]
@BastianZim 
Thanks.
I would like to add another package now, How would I do that?
I tried creating another pull request but since I already have the one for unifunc it seems I need to do it differently.

**Pull request merged**
